pkg/bootstrap/static.go

```diff
@@ -0,0 +1,392 @@
+						Address:       "127.0.0.1",
+						PortSpecifier: &core.SocketAddress_PortValue{PortValue: uint32(opts.AdminPort)},
+					},
+				},
+			},
+		},
+		StaticResources: &bootstrap.Bootstrap_StaticResources{
+			Clusters: BuildClusters(opts.Upstreams),
+		},
+	}
+
+	if opts.ZipkinAddress != "" {
+		zipkinAddress, err := toAddress(opts.ZipkinAddress)
+		if err != nil {
+			return nil, err
+		}
+		out.StaticResources.Clusters = append(out.StaticResources.Clusters,
+			v2.Cluster{
+				Name:           ZipkinCluster,
+				ConnectTimeout: 1 * time.Second,
+				Type:           v2.Cluster_STRICT_DNS,
+				LbPolicy:       v2.Cluster_ROUND_ROBIN,
+				Hosts:          []*core.Address{zipkinAddress},
+			})
+		out.Tracing = &trace.Tracing{
+			Http: &trace.Tracing_Http{
+				Name: util.Zipkin,
+				Config: toStruct(&trace.ZipkinConfig{
+					CollectorCluster:  ZipkinCluster,
+					CollectorEndpoint: "/api/v1/spans",
+				}),
+			},
+		}
+	}
+
+	// telemetry bootstrap should route telemetry directly to the local upstream
+	telemetryCluster := TelemetryCluster
+	if opts.TelemetryUpstream == nil {
+		telemetryAddress, err := toAddress(opts.TelemetryAddress)
+		if err != nil {
+			return nil, err
+		}
+		cluster := v2.Cluster{
+			Name:                 TelemetryCluster,
+			ConnectTimeout:       1 * time.Second,
+			Type:                 v2.Cluster_STRICT_DNS,
+			LbPolicy:             v2.Cluster_ROUND_ROBIN,
+			Hosts:                []*core.Address{telemetryAddress},
+			CircuitBreakers:      circuitBreakers,
+			Http2ProtocolOptions: &core.Http2ProtocolOptions{},
+		}
+		if opts.ControlPlaneAuth {
+			cluster.TlsContext = &auth.UpstreamTlsContext{
+				CommonTlsContext: commonTLSContext([]string{opts.TelemetrySAN}),
+			}
+		}
+		out.StaticResources.Clusters = append(out.StaticResources.Clusters, cluster)
+	} else {
+		telemetryCluster = clusterName(*opts.TelemetryUpstream)
+	}
+
+	out.StaticResources.Listeners = BuildListeners(opts.Upstreams, telemetryCluster, opts.ControlPlaneAuth)
+
+	if err := out.Validate(); err != nil {
+		return nil, err
+	}
+
+	return &out, nil
+}
+
+func commonTLSContext(altNames []string) *auth.CommonTlsContext {
+	return &auth.CommonTlsContext{
+		TlsCertificates: []*auth.TlsCertificate{{
+			CertificateChain: &core.DataSource{
+				Specifier: &core.DataSource_Filename{Filename: "/etc/certs/cert-chain.pem"},
+			},
+			PrivateKey: &core.DataSource{
+				Specifier: &core.DataSource_Filename{Filename: "/etc/certs/key.pem"},
+			},
+		}},
+		ValidationContext: &auth.CertificateValidationContext{
+			TrustedCa: &core.DataSource{
+				Specifier: &core.DataSource_Filename{Filename: "/etc/certs/root-cert.pem"},
+			},
+			VerifySubjectAltName: altNames,
+		},
+	}
+}
+
+func toAddress(address string) (*core.Address, error) {
+	host, portStr, err := net.SplitHostPort(address)
+	if err != nil {
+		return nil, err
+	}
+	port, err := strconv.Atoi(portStr)
+	if err != nil {
+		return nil, err
+	}
+	return &core.Address{
```

Comment:
This exists already as well (ip can be host too, we should update the param name): https://github.com/istio/istio/blob/97882cede142f5ac246a2e7cff4d6ff6bf631389/pilot/pkg/networking/util/util.go#L88